[Tablet Orders] Update product selection header and scanner button based on size class

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes #11748 

## Description
This PR updates the product selector header (left), along with the product scanner button within the order summary (right) so it's adapted for different size classes.

Since we've moved the scanner button from the header to the product selector for the moment, the only update to the header is the order in which we render the buttons and the search bar. So this PR mostly deals with the barcode scanner button.

![Simulator Screen Recording - iPhone 15 Plus - 2024-02-20 at 11 02 41](https://github.com/woocommerce/woocommerce-ios/assets/3812076/f4233212-60ff-4952-a244-157a50d43582)

## Changes
- Extracted parts of the `ProductSelectorView` logic into separate `@ViewBuilder` components: `productSelectorHeaderTitleRow`, and `productSelectorHeaderSearchRow`
- Duplicated the scanner logic from the current `scanProductButton` into a `scanProductRow` component. We could reuse the existing `scanProductButton` but this adds quite more complexity to a view that has already plenty of conditions. Since ideally we want to also decouple the scanner from the view entirely, I'd prefer to tackle both separately from this PR.
- When using `.sideBySide` presentation style, we'll present the updated header, as well as the `scanProductRow`
- When using `.modalOnModal` presentation style, we'll present the default header, as well as the default `scanProductButton`

## Testing instructions

### Case 1: Feature flag `false`
* With `sideBySideViewForOrderForm` set to `false`, the scanner button should work as always, that is: Along the `Add product" row when no products are in the order, and in the order summary header when products can be found in the order:
<img width="444" alt="Screenshot 2024-02-20 at 11 28 29" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/bf1147a4-e2f4-4f2f-99c8-1541fd5dffd5">
<img width="444" alt="Screenshot 2024-02-20 at 11 28 45" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/eb3c3d99-e00c-4abf-9ff3-c4735f6631c5">

### Case 2: Feature flag `true` - iPhone

| iPhone when no products in order | iPhone when products in order |
|--------|--------|
| <img width="298" alt="Screenshot 2024-02-20 at 11 19 07" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/59af4a49-be73-4849-8fdc-cb5be12e4a58"> | <img width="298" alt="Screenshot 2024-02-20 at 11 19 47" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/05ee1146-206d-49f5-9c8a-7e84db347db8"> | 

This case should work as always: 
* `[Plus icon] Add Products [Scanner icon]` is shown when no products are added to an order
* `Products [Scanner icon][Plus icon]` when products are added to an order

### Case 3: Feature flag `true` - iPhone split view

| iPhone when no products in order | iPhone when products in order |
|--------|--------|
| <img width="644" alt="Screenshot 2024-02-20 at 11 21 22" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/ac7ca99b-7884-44ee-92aa-1adca8ea7d04"> | <img width="640" alt="Screenshot 2024-02-20 at 11 22 22" src="https://github.com/woocommerce/woocommerce-ios/assets/3812076/6efa4f3f-8234-4d59-aba7-588c18f2690b"> |

This case should should both the default, and the updated UI: 
* Portrait: `[Plus icon] Add Products [Scanner icon]` is shown when no products are added to an order
* Portrait: `Products [Scanner icon][Plus icon]` when products are added to an order
* Landscape: Updated header
* Landscape: Updated scanner button

### Case 4: Feature flag `true` - iPad

| iPad when no products in order | iPhone when products in order |
|--------|--------|
| ![simulator_screenshot_07BEF6DF-4360-4C4C-87D1-4DDE051B11DE](https://github.com/woocommerce/woocommerce-ios/assets/3812076/b1f3afba-3cf7-4fc1-a91b-796e2c34cb35) | ![simulator_screenshot_DAB42310-6BF3-4D64-BA37-CEBD3F4ABF10](https://github.com/woocommerce/woocommerce-ios/assets/3812076/27b35b52-1b2b-4e32-b09c-22fd32217497) |

This case should show the updated UI.

iamgabrielma committed:
Extract product selector header into separate views

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
Merge branch 'trunk' into issue/11748-tablet-orders-m3-update-product-selection-header

iamgabrielma committed:
Remove scanner and grid button from header

iamgabrielma committed:
Restore scan button when products added to order

iamgabrielma committed:
Create scan button row. Render when sideBySide view

iamgabrielma committed:
Move scanProductRow. Update button style

iamgabrielma committed:
Apply IconButtonStyle

iamgabrielma committed:
Simplify button style

iamgabrielma committed:
Resolve incorrect background color on sidebyside

The view modifier was being applied only when .modalOnModal, making the row greyed-out on .sideBySide since the listForeground color was not being applied. This was done following the previous design update before deciding to use a full-width row for scan button.

iamgabrielma committed:
Revert adding withIconColor prop to IconButtonStyle

iamgabrielma committed:
Update vertical spacing if order has products

The way we current render and switch different components was adding additional padding to the scanner button when opening an order in a tablet, if no products are added to the order.

This happens because no other subviews are rendered inside the ProductsSection when there’s no products. By calculating vertical spacing dynamically we resolve this issue.

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift:
```diff
@@ -83,53 +83,14 @@ struct ProductSelectorView: View {
-                .buttonStyle(LinkButtonStyle())
-                .fixedSize()
+            if ServiceLocator.featureFlagService.isFeatureFlagEnabled(.sideBySideViewForOrderForm) &&
+                presentationStyle == .sideBySide {
+                productSelectorHeaderTitleRow
+                productSelectorHeaderSearchRow
+            } else {
+                productSelectorHeaderSearchRow
+                productSelectorHeaderTitleRow
```
I'm not sure if there's a specific design reason for having these two switched around in phone vs Tablet, but imho it seems like unnecessary complexity and creates an UX discrepancy. We should revisit it with Joe at some point and use the same layout in all devices.

joshheald replied:
I suspect it's because of the title role that `productSelectorHeaderTitleRow` fulfils in side-by-side mode ("Select products" or "2 products selected"), which it doesn't do in the modal presentation. But let's raise it in the design review request.

joshheald reviewed (commented):
A few visual issues to look at, outside this PR is fine:

1. A divider in the wrong place:
![divider position](https://github.com/woocommerce/woocommerce-ios/assets/2472348/629a1092-40dd-43fd-ab8f-3a989a6c30ee)

This looks like it's pre-existing, but we should log it as a design tweak at least.

2. Segmented control truncation in split view
![segmented control truncation](https://github.com/woocommerce/woocommerce-ios/assets/2472348/458de325-da7a-4961-9433-b55510acbd9a)

3. Header doesn't fit in split view
![header not fitting in split view](https://github.com/woocommerce/woocommerce-ios/assets/2472348/42c4b47a-6dbb-4e23-9b82-7b71af0dd926)

4. On iPhone, there's strange padding around the search bar
![padding-around-search](https://github.com/woocommerce/woocommerce-ios/assets/2472348/535af0cb-5c83-486f-adce-0eb68cc76022)

5. The search bar almost disappears when you use it on iPhone
![search-truncation-phone](https://github.com/woocommerce/woocommerce-ios/assets/2472348/8aea8156-59e9-4e6d-ab20-368f41b76b96)

6. The scan button purple is wrong for dark mode
![scan button color](https://github.com/woocommerce/woocommerce-ios/assets/2472348/e3d19cc9-2d79-47ba-92ee-f4346f357d41)

Otherwise working well – reviewing the code now.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift:
```diff
@@ -665,7 +677,58 @@ private struct ProductsSection: View {
+            case .notDetermined:
+                logPermissionStatus(status: .notDetermined)
+                viewModel.requestCameraAccess(onCompletion: { isPermissionGranted in
+                    if isPermissionGranted {
+                        showAddProductViaSKUScanner = true
+                        logPermissionStatus(status: .permitted)
+                    }
+                })
+            case .permitted:
+                showAddProductViaSKUScanner = true
+                logPermissionStatus(status: .permitted)
+            }
+        }, label: {
+            if showAddProductViaSKUScannerLoading {
+                ProgressView()
+            } else {
+                HStack() {
+                    Image(uiImage: .scanImage.withRenderingMode(.alwaysTemplate))
+                        .foregroundColor(Color(.brand))
+                    Text(Localization.scanProductRowTitle)
+                        .frame(maxWidth: .infinity, alignment: .leading)
+                }
+                .foregroundColor(Color(.accent))
+                .bodyStyle()
+            }
+        })
+        .frame(maxWidth: .infinity)
+        .sheet(isPresented: $showAddProductViaSKUScanner, onDismiss: {
+            scroll.scrollTo(addProductViaSKUScannerButton)
+        }, content: {
+            ProductSKUInputScannerView(onBarcodeScanned: { detectedBarcode in
+                showAddProductViaSKUScanner = false
+                showAddProductViaSKUScannerLoading = true
+                viewModel.addScannedProductToOrder(barcode: detectedBarcode, onCompletion: { _ in
+                    showAddProductViaSKUScannerLoading = false
+                }, onRetryRequested: {
+                    showAddProductViaSKUScanner = true
+                })
+            })
+        })
```
There's quite a lot of duplication from `scanProductButton` here... can that be avoided/reduced?

iamgabrielma replied:
You're right, I just went ahead with duplicating the logic before decoupling the whole thing, but there's no reason to keep it duplicated in this PR. I extracted the duplication into unique functions here: https://github.com/woocommerce/woocommerce-ios/pull/12033/commits/42a3b19694ff6a9d284c43a0da38edd84312932b

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/OrderForm.swift:
```diff
@@ -848,6 +911,13 @@ private extension ProductSelectorView.Configuration {
+    enum Localization {
+        static let scanProductRowTitle = NSLocalizedString(
+            "orderForm.products.add.scan.row.title",
+            value: "Scan product",
```
This should be title case: pecCkj-eD-p2
```suggestion
            value: "Scan Product",
```

iamgabrielma replied:
Fixed: https://github.com/woocommerce/woocommerce-ios/pull/12033/commits/910e4147e0902b14bcb0fda98831e12916c95b8f

joshheald reviewed (approved):
A string change, and question about reducing duplication, alongside the design issues in my earlier comment.

Feel free to break these out, and let me know if you need a hand with the conflict merge...

iamgabrielma committed:
Capitalize string

iamgabrielma committed:
Merge branch 'trunk' into issue/11748-tablet-orders-m3-update-product-selection-header

# Conflicts:
#	WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift

iamgabrielma committed:
Address non-existing variable from merge conflict

iamgabrielma committed:
Encapsulate scanner logic and content into methods

iamgabrielma commented:
> A few visual issues to look at, outside this PR is fine

Those are great, thanks Josh! I logged them here https://github.com/woocommerce/woocommerce-ios/issues/12057 to tackle them separately 🙇
